src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java

```diff
@@ -72,6 +82,45 @@ default boolean fullFillRecipe(@NotNull final IItemHandler... inventories)
      */
     boolean fullfillRecipe(final List<IItemHandler> handlers);
 
+    /**
+     * Get which type this recipe is
+     * CLASSIC or MULTI-OUTPUT are the only valid ones currently
+     * @return The recipe type
+     */
+    public RecipeStorageType getRecipeType();
+
+    /**
+     * Get a list of alternates to getPrimaryOutput
+     * @return a list if Itemstacks that this recipe can produce instead of getPrimaryOutput
+     */
+    public List<ItemStack> getAlternateOutputs();
+
+    /**
+     * Get the classic version of this recipe with GetPrimaryOutput targetted correctly from the chosen alternate
+     * @param requiredOutput Which output wanted
+     * @return the RecipeStorage that is "right" for that output
+     */
+    public RecipeStorage getClassicForMultiOutput(ItemStack requiredOutput);
+
+    /**
+     * Get the classic version of this recipe with GetPrimaryOutput targetted correctly from the chosen alternate
+     * @param stackPredicate Predicate to select the right stack
+     * @return the RecipeStorage that is "right" for that output
+     */
+    public RecipeStorage getClassicForMultiOutput(final Predicate<ItemStack> stackPredicate);
+
+    /**
+     * Source of the recipe, ie registry name.
+     * @return
+     */
+    public String getRecipeSource();
```

Comment:
If this is supposed to be a registry name, or reference one, Make it return a ResourceLocation instead,
If the return value is optional, mark it with `@Nullable` or return an optional and make the use of it like a functional invocation.

Reply:
I thought about that, I'd intended to support a variety of sources, including possibly player name. None of that is done yet though.

Reply:
I'll go ahead and convert to ResourceLocation, since there are difficulties in overloading it the way I planned with the way I ended up using it.